File: smartsite/src/types/worker.ts
/**
 * Core data contract for worker vitals from ESP32 sensors
 */
export interface WorkerVitals {
    /** Unique identifier for the worker (anonymized) */
    workerId: string;
    /** Heart rate in beats per minute */
    heartRate: number;
    /** Heart rate variability in milliseconds */
    hrv: number;
    /** Body temperature in Celsius */
    temperature: number;
    /** Number of sudden movements detected */
    jerkCount: number;
    /** Machine stress index (0-100) */
    machineStressIndex: number;
    /** Root mean square of vibration readings */
    vibrationRms: number;
    /** Construction Integrity Score (0-100) */
    cisScore: number;
    /** Risk state derived from cisScore thresholds */
    riskState: RiskState;
    /** Whether supervisor has issued a break flag */
    breakFlag: boolean;
    /** ISO timestamp of last sensor update */
    lastUpdated: string;
}

export type RiskState = "LOW" | "MEDIUM" | "HIGH";

/**
 * Risk style configuration for UI components
 */
export interface RiskStyleConfig {
    background: string;
    border: string;
    spotlight: string;
    textColor: string;
    badgeVariant: "default" | "secondary" | "destructive" | "outline";
}

/**
 * System status for dashboard header
 */
export type SystemStatus = "ONLINE" | "DEGRADED" | "OFFLINE";

/**
 * Dashboard statistics
 */
export interface DashboardStats {
    systemStatus: SystemStatus;
    totalActiveWorkers: number;
    highRiskCount: number;
    mediumRiskCount: number;
    lowRiskCount: number;
}

/**
 * Machine status types
 */
export type MachineStatus = "OPERATIONAL" | "WARNING" | "MAINTENANCE" | "OFFLINE";

/**
 * Machine data from IoT sensors for predictive maintenance
 */
export interface MachineData {
    /** Unique machine identifier */
    machineId: string;
    /** Machine display name */
    name: string;
    /** Machine type category */
    type: "CRANE" | "EXCAVATOR" | "LOADER" | "DRILL" | "COMPRESSOR" | "GENERATOR";
    /** Current operational status */
    status: MachineStatus;
    /** Stress index (0-100) */
    stressIndex: number;
    /** Operating temperature in Celsius */
    temperature: number;
    /** Vibration RMS value */
    vibrationRms: number;
    /** Operating hours today */
    operatingHours: number;
    /** Days since last service */
    daysSinceService: number;
    /** Fuel/power level percentage */
    fuelLevel: number;
    /** Oil pressure in PSI */
    oilPressure: number;
    /** Predicted failure probability (0-100) */
    failureProbability: number;
    /** Predicted days until maintenance needed */
    predictedMaintenanceDays: number;
    /** Health score (0-100) */
    healthScore: number;
    /** ISO timestamp of last update */
    lastUpdated: string;
}

File: backend/app/models/worker.py
"""
Worker and WorkerVitals database models
"""
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Integer, Float, Boolean, DateTime, ForeignKey, Enum, Uuid
from sqlalchemy.orm import relationship
import enum

from ..database import Base


class RiskState(str, enum.Enum):
    """Risk state enumeration"""
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class Worker(Base):
    """Worker registration table"""
    __tablename__ = "workers"
    
    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    worker_id = Column(String(20), unique=True, nullable=False, index=True)
    name = Column(String(100), nullable=True)
    department = Column(String(50), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    is_active = Column(Boolean, default=True)
    
    # Relationship to vitals
    vitals = relationship("WorkerVitals", back_populates="worker", order_by="desc(WorkerVitals.timestamp)")


class WorkerVitals(Base):
    """Worker vital signs from ESP32 sensors"""
    __tablename__ = "worker_vitals"
    
    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    worker_id = Column(String(20), ForeignKey("workers.worker_id"), nullable=False, index=True)
    
    # Vital signs
    heart_rate = Column(Integer, nullable=False)  # BPM
    hrv = Column(Integer, nullable=False)  # Heart rate variability in ms
    temperature = Column(Float, nullable=False)  # Body temperature in Celsius
    
    # Activity metrics
    jerk_count = Column(Integer, default=0)  # Sudden movements
    machine_stress_index = Column(Integer, default=0)  # 0-100
    vibration_rms = Column(Float, default=0.0)  # RMS vibration value
    
    # Computed scores
    cis_score = Column(Integer, nullable=False)  # Composite Index Score 0-100
    risk_state = Column(String(10), nullable=False)  # LOW, MEDIUM, HIGH
    
    # Flags
    break_flag = Column(Boolean, default=False)
    
    # Timestamp
    timestamp = Column(DateTime, default=datetime.utcnow, index=True)
    
    # Relationship
    worker = relationship("Worker", back_populates="vitals")
